functionary/prompt_utils.py

```diff
@@ -0,0 +1,210 @@
+    tokenizer: Any,
+    padding: str = "max_length",
+    max_length: Optional[int] = None,
+    return_tensor: bool = True,
+    verbose=False,
+) -> Tuple[str, Dict]:
+    """This function is used for when you want to get a dictionary input for the model.forward.
+    The dictionary will contain: input_ids, attention_maks, labels.
+    labels is like input_ids except that content from user, system, function will be set as -100, only content from assistant remains
+
+    Args:
+        messages (List[Dict]): List of messages in openAI format (containing: role, content and function_call (optional))
+        tokenizer (Any): tokenizer from transformers
+        padding (str, optional): type of padding (longest, max_length), this is passed to tokenizer(). Defaults to "max_length".
+        max_length (Optional[int], optional): maximum number of tokens allowed in prompt. Defaults to None.
+        return_tensor (bool, optional): if true, the input_dic will be dictionary[str, Tensor] else dictionary[str, List[int]]. Defaults to True.
+        verbose (bool, optional): to print some useful information or not. Defaults to False.
+
+    Returns:
+        Tuple[str, Dict]: (final_prompt_str, inputs)
+            final_prompt_str: the final prompt to be used,
+            inputs: a dictionary containing: input_ids, attention_mask, labels. This will be used in model.forward(**inputs)
+    """
+    # a dictionary mapping from token_id --> end_token
+    id_to_endtoken = get_token_id_to_end_token(tokenizer)
+    prompt_str = get_prompt_from_messages(messages)  # prompt_str is the concatenation of all prompts from messages
+    max_length = max_length if max_length is not None else tokenizer.model_max_length
+
+    input_dic = tokenizer(prompt_str, padding=padding, max_length=max_length, truncation=True)
+    input_token_ids = input_dic["input_ids"]
+    # first we initialize labels with all positions as -100,
+    # then we will fill in positions where role=assistant as we only include these in computing the loss
+    labels = [-100 for _ in range(len(input_token_ids))]
+    start = 0
+    # Now we find the positions where role=assistant and copy the value from input_token_ids
+    # this is done by finding the chunk (previous_stop_token_index + 1, current_stop_token_index + 1)
+    # where current_stop_token is EndToken.assistant or EndToken.function_call
+    for index, tok_id in enumerate(input_token_ids):
+        if tok_id in id_to_endtoken:  # Find position of end_token in final_prompt
+            end_token = id_to_endtoken[tok_id]
+            if (
+                end_token == EndToken.assistant or end_token == EndToken.function_call
+            ):  # only compute loss from tokens of assistant
+                for i in range(start + 2, index + 1):  # The reason for start + 2 is to ignore: "\nassistant" (2 tokens)
+                    labels[i] = input_token_ids[i]  # overwrite -100 to compute the loss
+                if verbose:
```

Comment:
We need to have tests for this instead of manually checking. For the whole prepare_training_inputs

Reply:
I have implemented unittest at: tests/test_prompt_creation.py